tests/conftest.py

```diff
@@ -25,3 +33,40 @@ def async_environment(request: typing.Any) -> str:
     ```
     """
     return request.param
+
+
+class ProxyWrapper(threading.Thread):
+    def __init__(self, host, port, **kwargs):
+        self.host = host
+        self.port = port
+        self.options = kwargs
+        super().__init__()
+
+    def run(self):
+        # mitmproxy uses asyncio internally but the default loop policy
+        # will only create event loops for the main thread, create one
+        # as part of the thread startup
+        asyncio.set_event_loop(asyncio.new_event_loop())
+        opts = options.Options(
+            listen_host=self.host, listen_port=self.port, **self.options
+        )
+        pconf = proxy.config.ProxyConfig(opts)
+
+        self.master = DumpMaster(opts)
+        self.master.server = proxy.server.ProxyServer(pconf)
+        self.master.run()
+
+    def join(self):
+        self.master.shutdown()
+        super().join()
+
+
+@pytest.fixture
+def proxy_server():
+    try:
+        thread = ProxyWrapper(PROXY_HOST, PROXY_PORT)
+        thread.start()
+        time.sleep(1)  # TODO: there's probably a better way to do this
```

Comment:
One quick dirty way might be to register an [addon](https://docs.mitmproxy.org/stable/addons-overview/) with the `.running()` callback…? [Masters trigger it just before running the mainloop](https://github.com/mitmproxy/mitmproxy/blob/e01f044c33aba51d9e5b6053e759b47594ad0855/mitmproxy/master.py#L77-L82). Eg:

```python
started = threading.Event()

class RunningNotify:
    def __init__(self):
        self.started = threading.Event()

    def running(self):
        self.started.set()

class ProxyWrapper:
    def __init__(self, ...):
        ...
        self.notify = RunningNotify()

    def run(self):
        ...
        master.addons.add(self.notify)
        ...

thread = ProxyWrapper(PROXY_HOST, PROXY_PORT)
thread.start()
thread.notify.started.wait()
```

Reply:
Nice find! I'll try it out, it's certainly dirty than the sleep 😄

Reply:
Worked like a charm, thanks!